### PR TITLE
SubstraTEE extension pack 1

### PR DESCRIPTION
**Pull request by brenzi**
## Grant Application

This application is (select one):
- [x] Speculative (use this by default)
- [ ] an RFP response

###

This application is (select one):
- [ ] Public (fully)
- [x] Public with private finances

### Abstract

SCS has developed SubstraTEE with a grant from the web3 foundation. After fulfilling that grant with the delivery of M4, SCS continued the development on its own and has reached M5, adding a private token transfer implementation that allows to instantiate a paint-balances module within an enclave. SubstraTEE has drawn the attention of the Energy Web Foundation and other industry contacts looking for confidentiality solutions for substrate or Blockchains in general. Also, synergies with other grant teams like pLibra or Advanca have been identified.

On request of Parity Technologies and W3F, SCS proposes an extension to SubstraTEE that aims at approaching a production-ready MVP solution for enterprise use cases.

## Checklist
- [x] The [grants](https://github.com/w3f/Web3-collaboration/blob/master/grants/grants.md) document has been read and understood.
- [x] The [Google Form](https://docs.google.com/forms/d/e/1FAIpQLSfMfjiRmDQDRk-4OhNASM6BAKii7rz_B1jWtbCPkUh6N7M2ww/viewform) will be completed accurately. Note that the Google Form requires the pull request URL.
- [x] Abstract (above) is succinct and complete.
- [x] The application is being included into the correct directory: either '[targeted](https://github.com/w3f/Web3-collaboration/tree/master/grants/targeted)' or '[speculative](https://github.com/w3f/Web3-collaboration/tree/master/grants/speculative)'.
- [x] The application includes a project description.
- [x] The application includes all names of team members.
- [x] The application includes a description of the team's experience.
- [x] The application includes all necessary links (e.g. GitHub and LinkedIn)
- [x] The "Development Roadmap" section in the application has a timeline of development ("milestones").
- [x] The "Development Roadmap" section in the application has an estimate of funds required.
- [x] The "Development Roadmap" section gives an indication of the team's long term plans.
- [x] The "Development Roadmap" section includes [documentation](https://github.com/w3f/Web3-collaboration/blob/master/grants/grants.md#documentation) as a deliverable for at least one milestone.

### CODE REVIEW

**Comment by brenzi** [comment]
@burdges Thanks for your very swift comments.

Regarding AMD: I'm afraid AMD (TrustZone) is no option just yet because they do not provide remote attestation. I'm actually putting more hope midterm into [KeyStone](https://keystone-enclave.org/)

Regarding session keys: I think managing session keys in an enclave could be done with reasonable effort. But I think if we'd like to really protect validators we should also consider to run the entire block generation and signing into the enclave. If we just protect keys within the enclave, how should we convince the enclave to sign a payload without risking exploits by malware? And if we are to move block generation into the enclave, this actual proposal is actually a requirement because we need trustless access to chain state for this. 

Regarding priorities: Our application answers a direct request from parity and people at web3. While I admit that TEEs like SGX are no "pure" confidentiality solution in the cryptographic sense, it seems that SGX is acceptable to many companies as an answer to GDPR requirements and to protect their trade secrets likewise.

**Comment by burdges** [comment]
No worries about TrustZone then.  


I'm speaking about validator nodes running primarily inside an enclave and attesting that their session keys cannot leave the enclave, sorry for the confusion.


Yes SGX surely helps with regulatory compliance somewhere, but..

SGX attestation cannot improve GDPR compliance for public blockchains, only zero-knowledge proofs achieve that.  Anyone who believe otherwise has poor legal advice and their data will end up being public.

We shall support non-public consortium chain designs on polkadot in which the block data remains secret, but only as second class citizens, so more like internal bridges than true parachains. 

Actual consortium "chains" still violate the deletion rules of the GDPR though, so either they must sacrifice being a blockchain by making old blocks unverifiable, or else they must nerf the definition of chain by using redactable or sanitizable signatures. 

Yes SGX might assist some consortium in running their collators, but that's roughly what I'm speaking about with the session keys comment.


In short, we cannot sell this project as providing either privacy or GDPR compliance.

**Comment by burdges** [comment]
Anyways we should fund this, apologies for dragging our own internal communications into your application.  :)

**Comment by brenzi** [comment]
I do actually appreciate to have this conversation publicly because I think it is important. While I agree with your reasoning from a theoretical point of view I'd state that both of us are not lawyers. Even worse, not even lawyers can give us dependable advice on what will be judged GDPR compliant in the blockchain space. While the [EU Blockchain Observatory has done great work](https://www.europarl.europa.eu/RegData/etudes/STUD/2019/634445/EPRS_STU(2019)634445_EN.pdf) in shedding some light at GDPR & Blockchain, there remain open questions.

I'm on thin ice here, but would claim that there is a good chance of SGX-solutions being considered as state-of-the-art for confidential computation. This way a vulnerability of SGX might not legally imply liability of the company that used this technology in good faith. However, that doesn't improve the fact that data may leak sooner or later.

But then, what is that data? In the case of SubstraTEE private-tx the data that is leaking is balances, not transaction history (given that we employ good forward secrecy by rotating state encryption keys)

regarding the right to erasure, SubstraTEE can provide exactly that: We don't keep history in the enclave. We keep state. We can delete and prune.

**Comment by brenzi** [comment]
Anyway, thanks for both criticism and endorsement ;-)

**Comment by burdges** [comment]
Ignoring GDPR, etc., there is a huge amount to be gained if we have the courage to trust TEEs with soundness aka financial security, instead of trusting it for privacy, but they both require similar development work.

As an extreme example, collators could simply threshold BLS blind sign coins and track nullifiers aka spent coins, so a distributed [Taler exchange](http://taler.net) (or other eCash mint).  If enough TEE enclaves get compromised then yes someone steals all the real tokens allotted by SPREEs to that parachain by signing unbacked tokens for themselves, but this requires breaking both the threshold assumption and the TEE at the same time, and it still does not destroy the underlying asset thanks to SPREEs.
